File: src/org/simbrain/util/math/DecayFunctions/LinearDecayFunction.java
package org.simbrain.util.math.DecayFunctions;

import org.simbrain.util.math.DecayFunction;
import org.simbrain.util.propertyeditor2.EditableObject;

public class LinearDecayFunction extends DecayFunction {

    @Override
    public double getScalingFactor(double distance) {
        double scalingFactor = 0;
        if (distance > getDispersion()) {
            return scalingFactor;
        } else {
            double stimulusDispersion = getDispersion();
            double peak = getPeakDistance();
            if (distance < peak) {
                scalingFactor = (stimulusDispersion - (2 * peak) + distance) / (stimulusDispersion - peak);

                if (scalingFactor < 0) {
                    scalingFactor = 0;
                }
            } else {
                scalingFactor = (stimulusDispersion - distance) / (stimulusDispersion - peak);
            }
            return scalingFactor;
        }
    }

    public static LinearDecayFunction create() {
        return new LinearDecayFunction();
    }

    public static LinearDecayFunctionBuilder builder() {
        return new LinearDecayFunctionBuilder();
    }

    @Override
    public EditableObject copy() {
        LinearDecayFunction ret = new LinearDecayFunction();
        ret.setDispersion(this.getDispersion());
        ret.setPeakDistance(this.getPeakDistance());
        ret.setRandomizer(this.getRandomizer().deepCopy());
        return ret;
    }

    public static class LinearDecayFunctionBuilder
            extends DecayFunctionBuilder<
            LinearDecayFunctionBuilder,
            LinearDecayFunction
            > {

        private LinearDecayFunction product = new LinearDecayFunction();

        @Override
        protected LinearDecayFunction product() {
            return product;
        }

        @Override
        public LinearDecayFunction build() {
            return product;
        }
    }
}

File: src/main/kotlin/org/simbrain/util/decayfunctions/LinearDecayFunction.kt
package org.simbrain.util.decayfunctions

class LinearDecayFunction @JvmOverloads constructor(dispersion: Double = 70.0) : DecayFunction(dispersion) {

    override fun getScalingFactor(distance: Double): Double {
        val dist = distanceFromPeak(distance)
        return if (dist > dispersion) 0.0 else baseMultiplier * (1 - dist / dispersion)
    }

    override fun copy(): LinearDecayFunction {
        return LinearDecayFunction(dispersion)
            .also {
                it.peakDistance = peakDistance
                it.baseMultiplier = baseMultiplier
            }
    }

    override val name = "Linear"

}